### PR TITLE
Mfbt - checkedInt - VPN-3315

### PR DESCRIPTION
**Pull request by bakulf**

### CODE REVIEW

File: src/mfbt/mozilla/CheckedInt.h [hunk 1]
```diff
@@ -0,0 +1,800 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
```
Comment (brizental) [inline src/mfbt/mozilla/CheckedInt.h:1]: Please add comments to this file and IntegerTypeTraits, mentioning they are copy pasted from m-c and linking to the copied files.

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp [hunk 2]
```diff
@@ -4,23 +4,24 @@
 
 #include "addonconditionwatchertimeend.h"
 #include "leakdetector.h"
+#include "mfbt/checkedint.h"
 #include "settingsholder.h"
 
 #include <QDateTime>
 
 AddonConditionWatcherTimeEnd::AddonConditionWatcherTimeEnd(QObject* parent,
```
Comment (brizental) [inline src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:12]: Is it possible to merge `AddonConditionWatcherTimeEnd` and `AddonConditionWatcherTimeStart` into a single class? They seem to be mostly redundant.

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp [hunk 3]
```diff
@@ -4,23 +4,24 @@
...
+  connect(&m_timer, &QTimer::timeout, this, [this]() {
+    if (maybeStartTimer()) {
+      emit conditionChanged(false);
+    }
+  });
 
-    connect(&m_timer, &QTimer::timeout, this,
-            [this]() { emit conditionChanged(false); });
-  }
+  maybeStartTimer();
```
Comment (brizental) [inline src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:24]: Don't we need to emit a `conditionChanged` if this returns `true`? What is the condition is already met when the condition watcher is started?

File: src/addons/conditionwatchers/addonconditionwatchertriggertimesecs.cpp [hunk 4]
```diff
@@ -20,21 +21,17 @@ AddonConditionWatcher* AddonConditionWatcherTriggerTimeSecs::maybeCreate(
 
 AddonConditionWatcherTriggerTimeSecs::AddonConditionWatcherTriggerTimeSecs(
```
Comment (brizental) [inline src/addons/conditionwatchers/addonconditionwatchertriggertimesecs.cpp:22]: This one is also redundant with the other time condition watchers.

File: src/models/devicemodel.h [hunk 5]
```diff
@@ -42,7 +42,7 @@ class DeviceModel final : public QAbstractListModel {
 
   void removeDeviceFromPublicKey(const QString& publicKey);
 
-  int activeDevices() const { return m_devices.count(); }
+  int activeDevices() const { return static_cast<int>(m_devices.count()); }
```
Comment (brizental) [inline src/models/devicemodel.h:45]: n00b question: Why do we need the static_cast here? Weren't we returning an `int` before already?

File: src/models/licensemodel.cpp [hunk 6]
```diff
@@ -60,8 +60,8 @@ void LicenseModel::initialize() {
 
   beginResetModel();
 
-  for (uint32_t i = 0, length = content.length(); i < length; ++i) {
-    int32_t lineLength = content[i].length();
+  for (qsizetype i = 0, length = content.length(); i < length; ++i) {
+    qsizetype lineLength = content[i].length();
```
Comment (brizental) [inline src/models/licensemodel.cpp:64]: Is this change actually related to the CheckedInt stuff or is it just a case of some implicit type coersion was going on before and is not anymore?

File: src/platforms/ios/iosiaphandler.h [hunk 7]
```diff
@@ -23,7 +23,7 @@ class IOSIAPHandler final : public PurchaseIAPHandler {
 
  protected:
   void nativeStartSubscription(ProductsHandler::Product* product) override;
-  void nativeRestoreSubscription();
+  void nativeRestoreSubscription() override;
```
Comment (brizental) [inline src/platforms/ios/iosiaphandler.h:26]: How is this change related to the chekedInt change?

File: tests/unit/testcheckedint.cpp [hunk 8]
```diff
@@ -0,0 +1,596 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (brizental) [inline tests/unit/testcheckedint.cpp:1]: If these are also copied from m-c, please also add a comment :)

**Review by brizental — COMMENTED** [review]
I left a few clarificaiton questions. Not sure I fully grasp what was done here yet :) Also, what does `mfbt` stand for?

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp (hunk 2 above)
Comment (bakulf) [reply src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:12]: follow up?

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp (hunk 3 above)
Comment (bakulf) [reply src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:24]: no signals in the CTOR. Nobody will receive them.

File: src/addons/conditionwatchers/addonconditionwatchertriggertimesecs.cpp (hunk 4 above)
Comment (bakulf) [reply src/addons/conditionwatchers/addonconditionwatchertriggertimesecs.cpp:22]: Why? this is not a specific time. It's X seconds from the first activation of the client.

File: src/models/devicemodel.h (hunk 5 above)
Comment (bakulf) [reply src/models/devicemodel.h:45]: count() returns a `qsizetype`, which means a `size_t`which is a `unsigned int`.

File: src/models/licensemodel.cpp (hunk 6 above)
Comment (bakulf) [reply src/models/licensemodel.cpp:64]: I removed a bunch of int warnings too.

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp (hunk 2 above)
Comment (brizental) [reply src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:12]: Sure!

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp (hunk 3 above)
Comment (brizental) [reply src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:24]: Ah, ok. But still what happens if the condition is already met when the condition watcher is started?

File: src/addons/conditionwatchers/addonconditionwatchertriggertimesecs.cpp (hunk 4 above)
Comment (brizental) [reply src/addons/conditionwatchers/addonconditionwatchertriggertimesecs.cpp:22]: Just because the new code looks like it's the same.

File: src/models/devicemodel.h (hunk 5 above)
Comment (brizental) [reply src/models/devicemodel.h:45]: Ok, but the function signature was not changed, which would mean it was already returning an `int` before. What that some implicit type coersion? What is different about that and the static_cast?

File: src/models/licensemodel.cpp (hunk 6 above)
Comment (brizental) [reply src/models/licensemodel.cpp:64]: Ah! Got it? Should we be failing CI on warning to avoid having things like these in the future?

File: src/platforms/ios/iosiaphandler.h (hunk 7 above)
Comment (brizental) [reply src/platforms/ios/iosiaphandler.h:26]: No need for a separate PR, but I have the same question as above https://github.com/mozilla-mobile/mozilla-vpn-client/pull/4979#discussion_r1029423504

File: src/models/devicemodel.h (hunk 5 above)
Comment (brizental) [reply src/models/devicemodel.h:45]: Nevermind, this was probably to address warnings. Right?

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp (hunk 3 above)
Comment (bakulf) [reply src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:24]: we call: `conditionApplied()` after the CTOR.

File: src/models/devicemodel.h (hunk 5 above)
Comment (bakulf) [reply src/models/devicemodel.h:45]: Right!

File: src/models/licensemodel.cpp (hunk 6 above)
Comment (bakulf) [reply src/models/licensemodel.cpp:64]: These warnings are shown only on iOS...

File: src/addons/conditionwatchers/addonconditionwatchertimeend.cpp (hunk 2 above)
Comment (bakulf) [reply src/addons/conditionwatchers/addonconditionwatchertimeend.cpp:12]: VPN-3333

File: src/models/licensemodel.cpp (hunk 6 above)
Comment (brizental) [reply src/models/licensemodel.cpp:64]: Sheesh, ok yeah let's forget about it now then 👀

**Review by brizental — APPROVED** [review]
